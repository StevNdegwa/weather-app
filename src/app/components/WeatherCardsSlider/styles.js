import { styled, Slider } from "@mui/material";
import { Slide } from "react-slideshow-image";

export const WeatheCardsSliderWrapper = styled(Slide)`
width: 95%;
background-color: none;
overflow: hidden;
border-radius: var(--border-radius-lg);
& button:disabled{
  opacity: 0;
}
`;

export const IndexSlider = styled(Slider)`
pointer-events: none;
& .MuiSlider-track {
  border: none;
  background-color: #90CBEC !important;
}
& .MuiSlider-rail {
  border: none;
  background-color: #DAE3E8 !important;
}
& .MuiSlider-thumb{
  display: none;
}
`;
