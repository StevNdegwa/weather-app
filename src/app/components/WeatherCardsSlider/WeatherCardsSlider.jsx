import {
  useCallback,
  useState,
  forwardRef,
  useLayoutEffect,
  useEffect,
} from "react";
import { IconButton } from "@mui/material";
import PropTypes from "prop-types";
import { GrNext, GrPrevious } from "react-icons/gr";
import "react-slideshow-image/dist/styles.css";
import WeatherForecastCard from "../WeatherForecastCard";
import { WeatheCardsSliderWrapper, IndexSlider } from "./styles";

const WeatherCardsSlider = forwardRef(
  ({ forecast = [], showTempChart }, ref) => {
    const [index, setIndex] = useState(0);
    const [slidesToShowCount, setSlidesToShowCount] = useState(1);

    const handleChange = useCallback(
      (oldIndex, newIndex) => {
        setIndex(newIndex);
      },
      [setIndex]
    );

    useLayoutEffect(() => {
      if (window.matchMedia) {
        if (window.matchMedia("(min-width: 800px)").matches) {
          setSlidesToShowCount(3);
        }
      }
    }, []);

    const slideOnKeyDown = useCallback(
      (event) => {
        if (event.code === "ArrowRight") {
          ref.current.goNext();
        } else if (event.code === "ArrowLeft") {
          ref.current.goBack();
        } else if (event.code === "Home") {
          ref.current.goTo(0);
        }
      },
      [ref]
    );

    useEffect(() => {
      window.addEventListener("keydown", slideOnKeyDown);
      return () => {
        window.removeEventListener("keydown", slideOnKeyDown);
      };
    }, [slideOnKeyDown]);

    const previousArrow = (
      <IconButton disabled={index === 0} aria-label="slide to previous card">
        <GrPrevious />
      </IconButton>
    );

    const nextArrow = (
      <IconButton
        disabled={index === forecast.length - 1}
        aria-label="slide to next card"
      >
        <GrNext />
      </IconButton>
    );

    return (
      <>
        <WeatheCardsSliderWrapper
          ref={ref}
          autoplay={false}
          transitionDuration={400}
          infinite={false}
          slidesToShow={slidesToShowCount}
          slidesToScroll={slidesToShowCount}
          onChange={handleChange}
          prevArrow={previousArrow}
          nextArrow={nextArrow}
        >
          {forecast.map(({ description, date, icon, temperature }, index) => (
            <WeatherForecastCard
              key={index}
              date={date}
              description={description}
              icon={icon}
              temperature={temperature}
              showTempChart={showTempChart}
            />
          ))}
        </WeatheCardsSliderWrapper>
        <IndexSlider
          aria-label="Index"
          value={index}
          readOnly
          min={0}
          max={forecast.length - slidesToShowCount}
          tabIndex={-1}
        />
      </>
    );
  }
);

WeatherCardsSlider.propTypes = {
  forecast: PropTypes.array,
  showTempChart: PropTypes.func.isRequired,
};

export default WeatherCardsSlider;
