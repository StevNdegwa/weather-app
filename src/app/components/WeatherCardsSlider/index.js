import WeatherCardsSlider from "./WeatherCardsSlider";
export default WeatherCardsSlider;
