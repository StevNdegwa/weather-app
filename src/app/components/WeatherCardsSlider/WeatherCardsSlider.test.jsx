import { createRef } from "react";
import { mount } from "enzyme";
import { Slide } from "react-slideshow-image";
import { Slider } from "@mui/material";
import { act } from "@testing-library/react";
import WeatherCardsSlider from "./WeatherCardsSlider";

describe("test <WeatherCardsSlider/>", () => {
  let wrapper,
    forecast = [],
    showTempChart = jest.fn(),
    ref = createRef();

  beforeEach(() => {
    wrapper = mount(
      <WeatherCardsSlider
        forecast={forecast}
        showTempChart={showTempChart}
        ref={ref}
      />
    );
  });

  test("snapshot testing", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should render right components", () => {
    expect(wrapper.find(Slide)).toHaveLength(1);
    expect(wrapper.find(Slider)).toHaveLength(1);
  });

  test("should update slide index", async () => {
    expect(wrapper.find(Slider).prop("value")).toEqual(0);

    await act(async () => {
      wrapper.find(Slide).prop("onChange")(0, 1);
    });

    wrapper.update();

    expect(wrapper.find(Slider).prop("value")).toEqual(1);

  });
});
