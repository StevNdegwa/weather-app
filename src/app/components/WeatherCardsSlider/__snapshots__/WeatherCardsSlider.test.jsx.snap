// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <WeatherCardsSlider/> snapshot testing 1`] = `"<div dir=\\"ltr\\" aria-roledescription=\\"carousel\\" class=\\"css-1ohr7kt\\"><div class=\\"react-slideshow-container\\"><button class=\\"MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium undefined nav disabled css-78trlr-MuiButtonBase-root-MuiIconButton-root\\" tabindex=\\"-1\\" type=\\"button\\" disabled=\\"\\" aria-label=\\"Previous Slide\\" data-type=\\"prev\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 24 24\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" points=\\"7 2 17 12 7 22\\" transform=\\"matrix(-1 0 0 1 24 0)\\"></polyline></svg></button><div class=\\"react-slideshow-wrapper slide \\"><div class=\\"images-wrap\\" style=\\"transform: translate(-0px); width: 0px;\\"></div></div><button class=\\"MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium undefined nav disabled css-78trlr-MuiButtonBase-root-MuiIconButton-root\\" tabindex=\\"-1\\" type=\\"button\\" disabled=\\"\\" aria-label=\\"Next Slide\\" data-type=\\"next\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 24 24\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" points=\\"7 2 17 12 7 22\\"></polyline></svg></button></div></div><span readonly=\\"\\" class=\\"MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-10yivjf-MuiSlider-root\\"><span class=\\"MuiSlider-rail css-14pt78w-MuiSlider-rail\\"></span><span class=\\"MuiSlider-track css-1gv0vcd-MuiSlider-track\\" style=\\"left: 0%; width: 100%;\\"></span><span data-index=\\"0\\" class=\\"MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb\\" style=\\"left: 100%;\\"><input tabindex=\\"-1\\" data-index=\\"0\\" aria-label=\\"Index\\" aria-orientation=\\"horizontal\\" aria-valuemax=\\"-1\\" aria-valuemin=\\"0\\" aria-valuenow=\\"-1\\" type=\\"range\\" min=\\"0\\" max=\\"-1\\" style=\\"border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;\\" value=\\"-1\\"></span></span>"`;
